### PR TITLE
Update webpack to version 4, babel to version 7

### PR DESCRIPTION
**Pull request by hstastna**
Upgrade the project's webpack to version 4, and babel to version 7.  Webpack configurations and plugins updated as necessary to support the upgrade.  Babel translation and polyfills updated as necessary.

Webpack maintains the unique dev and prod config files.  Babel is configured with a pair of project presets.  These allow babel to be updated independently from webpack's configs.  Specific configurations (optimizations plugins, general plugins, file naming, babel transform set, jest configurations) have been adapted to the project from ovirt-engine-ui-extensions as well as an ejected create react app sample application (https://create-react-app.dev/docs/getting-started/).

This PR includes the required changes of the webpack config files _webpack.config.dev.js_, and _webpack.config.prod.js_.

Some other packages/plugins were updated:
- _babel-jest_
- _case-sensitive-paths-webpack-plugin_
- _copy-webpack-plugin_
- _flow-babel-webpack-plugin_
- _handlebars_
- _html-webpack-plugin_
- _prop-types_
- _webpack-dev-server_

Updated loaders:
- _babel-loader_
- _css-loader_
- _file-loader_
- _handlebars-loader_
- _html-loader_
- _postcss-loader_
- _style-loader_
- _url-loader_

Some deprecated packages replaced wit the new ones:
- _babel-core_ replaced with _@babel/core_
- _babel-preset-latest_ replaced with _@babel/preset-env_
- _babel-preset-react_ replaced with _@babel/preset-react_
- _eslint-loader_ replaced with _eslint-webpack-plugin_
- _extract-text-webpack-plugin_ replaced with _mini-css-extract-plugin_

Also `devtool` was set to `'eval-source-map'`, same as in the _ovirt-engine-ui-extensions_.

### CODE REVIEW

File: config/webpack.config.prod.js [hunk 1]
```diff
@@ -59,12 +59,17 @@ module.exports = {
     // We use `fallback` instead of `root` because we want `node_modules` to "win"
     // if there any conflicts. This matches Node resolution mechanism.
     // https://github.com/facebookincubator/create-react-app/issues/253
-    fallback: paths.nodePaths,
+    // Note: resolve.fallback and some others were replaced by a single option resolve.modules
+    modules: [
```
Comment (sjd78) [inline config/webpack.config.prod.js:63]: Like dev, I don't think we need this block

File: config/webpack.config.prod.js [hunk 2]
```diff
@@ -76,28 +81,31 @@ module.exports = {
   // directory of `react-scripts` itself rather than the project directory.
   // You can remove this after ejecting.
   resolveLoader: {
-    root: paths.ownNodeModules,
-    moduleTemplates: ['*-loader'],
+    // TODO (hstastna): what to do with this? it does not work for webpack > 1
```
Comment (sjd78) [inline config/webpack.config.prod.js:84]: You can remove the `resolveLoader` block all together

File: config/webpack.config.prod.js [hunk 3]
```diff
@@ -76,28 +81,31 @@ module.exports = {
...
+        // First, run the linter.
+        // It's important to do this before Babel processes the JS.
         test: /\.(js|jsx)$/,
-        loader: 'eslint',
+        loader: 'eslint-loader',
         include: paths.appSrc,
         options: {
           failOnError: true,
-        }
+        },
```
Comment (sjd78) [inline config/webpack.config.prod.js:101]: I think you can add `configFile: path.join(__dirname, 'eslint.js'),` here and be ok and not need it later in the `LoaderOptiosnsPlugin`

File: config/webpack.config.prod.js [hunk 4]
```diff
@@ -123,26 +131,48 @@ module.exports = {
...
+          }
+        ]
       },
+
       // JSON is not enabled by default in Webpack but both Node and Browserify
       // allow it implicitly so we also enable it.
       {
         test: /\.json$/,
-        loader: 'json',
+        loader: 'json-loader',
```
Comment (sjd78) [inline config/webpack.config.prod.js:167]: I don't think this loader is needed anymore

File: config/webpack.config.dev.js [hunk 5]
```diff
@@ -224,6 +227,31 @@ module.exports = {
...
+                'Firefox ESR',
+                'not ie > 0', // VM Portal don't support IE at all
+              ],
+              grid: false,
+            }),
+          ]
+        },
+        // Point ESLint to our predefined config.
+        eslintLoader: {
+          configFile: path.join(__dirname, 'eslint.js'),
```
Comment (sjd78) [inline config/webpack.config.dev.js:249]: I'm fairly sure this can go up in the rules.[].options instead of being here.

**Review by sjd78 — DISMISSED** [review]
I didn't get through the whole thing yet, I'm still testing it out.  First round of comments:

In general, I think a lot of the comments can be cut down or removed since they're only partially applicable anymore.

On dev - `devtool` can be set to `eval-source-map`.  That setting works well in ui-extensions.

I think the json-loader can be dropped (https://v4.webpack.js.org/loaders/json-loader/)

File: config/webpack.config.prod.js (hunk 4 above)
Comment (hstastna) [reply config/webpack.config.prod.js:167]: Right. I can see I removed that in _config/webpack.config.dev.js_ and forgot to do it also here. Thank you! 👍🏽

File: package.json [hunk 6]
```diff
@@ -84,7 +84,7 @@
     "connected-react-router": "4.3.0",
     "history": "4.7.2",
     "immer": "7.0.7",
-    "immutable": "3.8.2",
+    "immutable": "4.0.0-rc.12",
```
[the remark this answers is not in the document]
Comment (hstastna) [reply package.json:87]: I am not aware of any breaking changes. This version bump just solved some issues I had.

File: config/webpack.config.dev.js [hunk 7]
```diff
@@ -63,12 +63,17 @@ module.exports = {
     // We use `fallback` instead of `root` because we want `node_modules` to "win"
     // if there any conflicts. This matches Node resolution mechanism.
     // https://github.com/facebookincubator/create-react-app/issues/253
-    fallback: paths.nodePaths,
+    // Note: resolve.fallback and some others were replaced by a single option resolve.modules
+    modules: [
```
[the remark this answers is not in the document]
Comment (hstastna) [reply config/webpack.config.dev.js:67]: Hmm, are you sure? I think this is quite default. Do you suggest completely to remove this "modules" part?

**Comment by sgratch** [comment]
when trying to verify then login crashed with the following error:
image

**Comment by rszwajko** [comment]
When running dev mode I still see Flow errors about incorrect type definitions in Immutable.js library. It would be nice to refresh WebPack <-> Babel <-> Flow integration and get rid of them. 
From what I could check:
1. flow-babel-webpack-plugin seems [dead project](https://github.com/zhirzh/flow-babel-webpack-plugin)  (if I correctly traced the origin of the plugin)
2. we don't use `@babel/preset-flow` in our WebPack config which seems the [recommended way](https://flow.org/en/docs/install/)

**Comment by rszwajko** [comment]
In the current version dev server fails on my env with:
```
(node:13843) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
(node:13843) UnhandledPromiseRejectionWarning: TypeError: missingDeps.some is not a function
    at ovirt-web-ui/scripts/utils/WatchMissingNodeModulesPlugin.js:15:21
    at _next0 (eval at create (ovirt-web-ui/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:8:1)
    at eval (eval at create (ovirt-web-ui/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:21:1)
    at ovirt-web-ui/node_modules/html-webpack-plugin/index.js:303:11
(node:13843) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:13843) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Review by sjd78 — COMMENTED** [review]
Looking at everything in the PR so far and:
  - dev server works
  - webpack is v4
  - babel is v7
  - eslint is a manual process (outside of the dev/prod build chain)
  - flow is a manual process (outside of the dev/prod build chain)'
  - production build configs need to be updated
  - CI fails on travis because prod build and jest configs needs to be updated
  - CI fails on std-ci because prod build and jest configs needs to be updated and nodejs-modules need a pre-seed

To get to a minimal upgrade level:
- [x] Ensure eslint can be run against app sources manually
- [x] Ensure flow can be run against app sources manually
- [x] Ensure dev server starts and the app runs
- [ ] Ensure production build completes and that when installed to an ovirt-engine it runs (including branding)
- [ ] Pre-seed nodejs-modules once prod build is verified working

**Comment by sjd78** [comment]
To get to a minimal upgrade level:
- [x] Ensure eslint can be run against app sources manually
- [x] Ensure flow can be run against app sources manually
- [x] Ensure dev server starts and the app runs
- [x] Ensure jest tests run and complete successfully
- [x] Ensure production build completes
- [x] Ensure production build works when installed to an ovirt-engine
- [ ] Ensure production build works with RHV branding when installed on ovirt-engine
- [ ] Pre-seed nodejs-modules once prod build is verified working

**Comment by sjd78** [comment]
(rebased)

**Comment by sjd78** [comment]
ci test

**Comment by sjd78** [comment]
ci test

**Comment by sjd78** [comment]
There are some changes to be done in the future to compliment this work.  In the current state, this PR gets us to webpack 4 and babel 7 at a functional level for both the dev server and for the production build.

Future work in no particular priority order:

 - Upgrade and integrate eslint to the webpack build.  The upgrade will give us better linting.  Integrating with the webpack build will show lint errors during development and optionally prevent the prod/CI build from succeeding with lint errors in place.

 - Convert from flow to typescript.  This will give us better typing support in builds and tooling.  Also, gives us the option to drop react PropTypes in favor of normal typescript typing.  Patternfly-react has been doing this for awhile now.  There are a number of ways to automatically convert flow to typescript.  For an example: https://skovhus.github.io/blog/flow-to-typescript-migration/

 - Figure out what settings and code needs to be tweaked to run the production build in `mode: 'production'` instead of `mode: 'development'`.  There are tree-shaking and other optimizations that are automatically turned on with the production mode that caused problems with the app.  I already ruled out the `TerserPlugin` and the `OptimizeCSSAssetsPlugin` webpack plugins as the source of the error.  I'm confident the error is not in the babel transform layer.  Unfortunately, working on this problem requires build and installing the prod build on a dev ovirt-engine to properly debug and work on.

 - Postcss may not be all that useful for us so may be removed to save build time as css file size.

 - File sizes of the built code are reported in gzip size.  When deployed to production, I'm not sure a gzip filter is installed/enabled on the ovirt-web-ui.war.  We may be able to make life easier for users with high latency by setting up gzip filters on the static content.

 - Evaluate the docker build (see issue #1442)

**Comment by rszwajko** [comment]
>  * Figure out what settings and code needs to be tweaked to run the production build in `mode: 'production'` instead of `mode: 'development'`.  

It seems that the production build works for me with following changes:
```
diff --git a/config/webpack.config.prod.js b/config/webpack.config.prod.js
index 1f540c1..498f51c 100644
--- a/config/webpack.config.prod.js
+++ b/config/webpack.config.prod.js
@@ -44,7 +44,7 @@ module.exports = ((webpackEnv) => {
   let fontsToEmbed
 
   const theConfig = {
-    mode: 'development', //'production',
+    mode: 'production',
     bail: true,
     devtool: isEnvDevelopment ? 'eval-source-map' : 'source-map',
 
diff --git a/package.json b/package.json
index a146b02..b741af1 100644
--- a/package.json
+++ b/package.json
@@ -86,9 +86,9 @@
     "lodash": "4.17.21",
     "moment": "2.27.0",
     "moment-duration-format": "2.3.2",
-    "patternfly": "3.59.3",
+    "patternfly": "3.59.5",
     "patternfly-bootstrap-combobox": "1.1.7",
-    "patternfly-react": "2.36.9",
+    "patternfly-react": "2.39.17",
     "popper.js": "1.14.3",
     "prop-types": "15.7.2",
     "react": "16.8.6",

```

**Comment by sjd78** [comment]
@hstastna , @rszwajko , @sgratch - Radek's patternfly and patterfly-react tweaks works well.  The production build runs in production mode and the app works without further changes.

**Comment by sjd78** [comment]
ci test

**Comment by sjd78** [comment]
Rebased and because of @rszwajko, then snapshot info popup is working again:

VM Portal - Google Chrome_034

**Pull request merged**